src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Windows.targets

```diff
@@ -84,19 +84,19 @@ The .NET Foundation licenses this file to you under the MIT license.
     </ItemGroup>
 
     <ItemGroup>
-      <LinkerArg Condition="'$(NativeLib)' == 'Shared'" Include="/DLL" />
+      <LinkerArg Condition="'$(NativeLib)' == 'Shared' or '$(CustomNativeMain)' == 'true'" Include="/DLL" />
       <LinkerArg Include="@(NativeLibrary->'&quot;%(Identity)&quot;')" />
       <LinkerArg Include="@(SdkNativeLibrary->'&quot;%(Identity)&quot;')" />
       <LinkerArg Include="/NOLOGO /MANIFEST:NO" />
       <LinkerArg Condition="$(NativeDebugSymbols) == 'true'" Include="/DEBUG" />
       <!-- The runtime is not compatible with jump stubs inserted by incremental linking. -->
       <LinkerArg Include="/INCREMENTAL:NO" />
-      <LinkerArg Condition="'$(LinkerSubsystem)' != ''" Include="/SUBSYSTEM:$(LinkerSubsystem)" />
-      <LinkerArg Condition="'$(OutputType)' == 'WinExe' or '$(OutputType)' == 'Exe'" Include="/ENTRY:$(EntryPointSymbol) /NOEXP /NOIMPLIB" />
+      <LinkerArg Condition="'$(LinkerSubsystem)' != '' and '$(CustomNativeMain)' != 'true'" Include="/SUBSYSTEM:$(LinkerSubsystem)" />
+      <LinkerArg Condition="('$(OutputType)' == 'WinExe' or '$(OutputType)' == 'Exe') and '$(CustomNativeMain)' != 'true'" Include="/ENTRY:$(EntryPointSymbol) /NOEXP /NOIMPLIB" />
```

Comment:
@jkotas another question (I'm not too familiar with how this works here), do we need to also pass `/EXPORT:$(EntryPointSymbol)` in this scenario to make the .dll export the custom entry point, or is that done by default?

Reply:
Another option may be to set `IlcExportUnmanagedEntrypoints` in CsWinRT SDK. It would apply to all named UnmanagedCallersOnly method.

Either setting `IlcExportUnmanagedEntrypoints` to true or adding `/EXPORT` for the few entrypoints that you care about would work. The solutions differ in the level of additional customization. `IlcExportUnmanagedEntrypoints` would not allow people to use named `UnmanagedCallersOnly` as internal implementation detail that they do not want to expose.

Adding `/EXPORT:__managed__Main` and `/EXPORT:DllGetActivationFactory` to `LinkerArgs` in CsWinRT SDK is probably better.

Reply:
I see, yeah we can probably do that! Using `/EXPORT` would also be properly supported I assume, whereas `IlcExportUnmanagedEntrypoints` would be an undocumented NativeAOT option (due to the "Ilc-" prefix), which we likely wouldn't want to take a dependency on in stable CsWinRT.

So to recap, our .targets in CsWinRT should be something like this (under some CsWinRT-specific property):
```xml
<PropertyGroup>
  <CustomNativeMain>true</CustomNativeMain>
  <NativeLib>Shared</NativeLib>
  <EntryPointSymbol>CsWinRT_StubExeEntryPoint</EntryPointSymbol>
</PropertyGroup>
<ItemGroup>
  <LinkerArg Include="/EXPORT:$(EntryPointSymbol)" />
  <LinkerArg Include="/EXPORT:DllGetActivationFactory" />
  <LinkerArg Include="/EXPORT:DllCanUnloadNow" />
</ItemGroup>
```

This would be all we need there (other than the stub .exe itself), right?

Reply:
How would be `CsWinRT_StubExeEntryPoint` implemented?

I think `EntryPointSymbol` may want to be `__managed__Main` unless you need some additional customizations. It is unmanaged entrypoint generated by the AOT toolchain. This name is hardcoded in the toolchain. (iOS SDK has a dependency on it as well: https://github.com/search?q=repo%3Axamarin%2Fxamarin-macios%20__managed__Main&type=code.)

Reply:
I think that's the part I was missing, it wasn't clear to me whether the toolchain would be respecting `EntryPointSymbol` and using that name to generate the custom entry point for the binary, or something else. I can see that's not the case then. I assume since iOS also depends on it, it's fine for us in CsWinRT to also take a dependency on it?

Reply:
Right.